### ISSUE TITLE
Error band on ellipse

### ISSUE THREAD

**Issue by evanberkowitz** [evanberkowitz/lsqfit-examples#1]
In `ellipse.py` I generated some data by taking some points on an ellipse, displacing them randomly, and assigning random error bars, the aim being to recover the input parameters of the ellipse (displacement, tilt, minor and major axes).

Because an ellipse is X-to-many-Y, it actually makes sense to shift perspective and think of (X,Y) as the independent data and the "fulfills the ellipse's equation" as the dependent data.  That is, given the constraint

```
    g(X, Y) = 
            [ (x-x0) * cos phi + (y-y0) * sin phi ]^2 / a^2 + 
            [ (x-x0) * sin phi - (y-y0) * cos phi ]^2 / b^2 
            - 1
```

the data points all claim to satisfy `g(X, Y) = 0` (with an exact 0).  So, this case has all the errors on the data on the x variables (so the data should be put into the priors, according to https://lsqfit.readthedocs.io/en/latest/overview.html#x-has-errors) and the y has NO errors (https://lsqfit.readthedocs.io/en/latest/overview.html#y-has-no-error-marginalization); I just give extremely small errors `0.0(1.0)e-5`.

I get a very nice fit, shown below.

![ellipse](https://user-images.githubusercontent.com/129339/87983676-bf3c3f80-caa6-11ea-9bc8-2a55285765a3.png)

But the issue is---I would like to see an error band on the best-fit ellipse.  I would expect, given the data, that error bands that account for covariances for should more poorly constrain the ellipse in the lower-left corner.

This seems to be quite a difficult, or at least non-intuitive thing to accomplish.  Perhaps @gplepage has an idea or a hint as to how to get this to work?

**Comment by evanberkowitz** [comment]
In the changes up to and including 7d45715 I made some changes that I thought would reliably give me the correct bands.

Plugging in the best fit, I am left with a function that produces a parametric description of the ellipse `r(theta)`.  Running theta from 0 to 2π and getting `r.sdev`, and adding/subtracting from `r.mean` in to get 1- and 2-sigma bands yields

![ellipses](https://user-images.githubusercontent.com/129339/87987907-67550700-caad-11ea-9890-637a47efc594.png)

black is the original ellipse, blue is the best-fit, orange and green are 1- and 2-sigma "bands".

I am happy, in that my expectation that the uncertainty in the lower-left is much greater than in other places (due to the lack of data there).  But in some places the original ellipse is far outside the 2-sigma bands.  This makes me think that I am omitting some contribution to the uncertainty.

(note that the error bars changed because I `np.sort`ed the angles, to make it easier to scan the list of data.)

**Comment by evanberkowitz** [comment]
I must have a bug.  I cranked up to 100 points, put the displacements to 0 (so that every central value is perfect), and shrank the error bars.  And yet, I find

![ellipse-many-points](https://user-images.githubusercontent.com/129339/87989834-bfd9d380-cab0-11ea-9344-fb63a785634c.png)

with a fit of

```
Least Square Fit:
  chi2/dof [dof] = 0.0028 [100]    Q = 1    logGBF = 263.47

Parameters:
             x0   10.0000 (27)      [   10.0 (3.0) ]
             y0   12.0000 (17)      [   12.0 (2.0) ]
            phi   0.24309 (91)      [    0.0 (3.0) ]
              a    3.0277 (31)      [    2.0 (2.0) ]
              b    0.9126 (19)      [    1.0 (1.0) ]
```

The fact that a, b, and phi are all so many sigma from their true values implies that I must have an inconsistency between the meaning of the different parameters in my exact forms.

**Comment by evanberkowitz** [comment]
OK, there was a mistaken sign in the ellipse function.  Now the exact / fit comparison looks like

![ellipse-fixed](https://user-images.githubusercontent.com/129339/87992644-b2274c80-cab6-11ea-8643-832d6a3f1645.png)

which is, needless to say, substantially better.  For many angles, the ellipses are nearly on top of one another; they get as far apart as about 1σ in the lower-left.

The fit result is
```
Least Square Fit:
  chi2/dof [dof] = 0.56 [20]    Q = 0.94    logGBF = -6.5944

Parameters:
             x0     9.85 (14)     [  10.0 (3.0) ]
             y0   11.926 (91)     [  12.0 (2.0) ]
            phi   -7.613 (38)     [   0.0 (3.0) ]  **
              a    1.063 (65)     [   2.0 (2.0) ]
              b     3.15 (18)     [   2.0 (2.0) ]
```

This is pessimistic because phi is only defined up to π/2, if you're willing to switch a with b.  The ellipse was seeded with `a=3, b=1, phi=0.2`.  But here we get `a~=1, b~=3, phi~=0.241(38) (mod π/2)`.  So it's got the right fit, but the prior of phi between 0±π wasn't enforced and the chi2/dof is mistaken; adjusting the prior to be `0(π/4 = 0.78)` does the trick:

```
Least Square Fit:
  chi2/dof [dof] = 0.25 [20]    Q = 1    logGBF = -2.1509

Parameters:
             x0     9.90 (12)     [  10.0 (3.0) ]
             y0   11.951 (84)     [  12.0 (2.0) ]
            phi    0.230 (38)     [   0.00 (78) ]
              a     3.10 (15)     [   2.0 (2.0) ]
              b    1.060 (65)     [   2.0 (2.0) ]
```

**Comment by evanberkowitz** [comment]
Still, the question remains:  are the lines I've drawn really the right error bands?  They seem to get very narrow around `(7.5, 12.1)`; is that right?  It is apparent that 180˚ away are a lot of points which cause a dramatic narrowing of the error bands as well.  So it might be a sensible "echo".

**Comment by gplepage** [comment]
I responded with email, answering your question. Just want to add one plot which shows the true 1-sigma error band (taking account of correlations between the ``x`` and ``y`` uncertainties). The error band narrows around ``(7.1, 11.6).`` This is because of the one data point there with an unusually small x-error. 
![Figure_1](https://user-images.githubusercontent.com/1808545/88194883-54f1df00-cc0d-11ea-91ae-a240e2eaee2b.png)

Sorry that was for different data (from your seed multiplied by 4). This is the plot that goes with the data above. It narrows around ``(12.9, 12.4)`` because there is so much data concentrated there (so the cumulative error is smaller).

![Figure_1](https://user-images.githubusercontent.com/1808545/88196744-8a97c780-cc0f-11ea-9ce6-a0c8debc6f0a.png)

**Comment by evanberkowitz** [comment]
In `gvar_plot.py` I give some utilities for plotting gvars, including a utility that, given an x and y, plots error ellipses around it.  I updated `ellipse.py` to use that utility on lots of plugged-in best-fit gvars, and get fully-correlated bands

![zillions](https://user-images.githubusercontent.com/129339/88358532-4efb1b80-cd3d-11ea-94af-aaa115c852d4.png)

which looks very similar to the plot from @gplepage.  In my plot one can see sharp lines for 1, 2, and 3-sigma bands.  I think in places where it's more saturated the error ellipses tend to be aligned with the tangent of the best fit, while in places where it's less saturated the error ellipses tend to be more normal to the best fit.

This resolves #1.  However, I think my `gvar_plot.py` is nice enough and bare-bones enough that it could be included in `gvar` itself as `gvar.plot`.  It would be nice, eventually, to support `gvar.{PowerSeries,CSpline,Pade}` objects.

**Comment by evanberkowitz** [comment]
Actually, I think the sigma multiplier goes _outside_ the sqrt.  So the actual 1,2,3-σ bands look like

![zillions](https://user-images.githubusercontent.com/129339/88406545-d462e800-cd9e-11ea-8171-bc6d72326acd.png)

which I think looks more believable.
